lib/pos_mempool.go

```diff
@@ -0,0 +1,397 @@
+	// The first fee range has a fee rate of feeBucketBaseRateNanos.
+	if index == 0 {
+		fee, _ := feeBucketBaseRateNanos.Uint64()
+		return fee
+	}
+
+	// Compute feeBucketBaseRateNanos * feeBucketMultiplier^index.
+	pow := NewFloat().SetInt(big.NewInt(int64(index)))
+	multiplier := BigFloatPow(feeBucketMultiplier, pow)
+	fee := NewFloat().Mul(feeBucketBaseRateNanos, multiplier)
+
+	feeUint64, _ := fee.Uint64()
+	return feeUint64
+}
+
+// ComputeTimeBucketRangeFromIndex takes a fee range index, feeBucketBaseRateNanos, and feeBucketMultiplier, and returns the
+// [minFeeNanosPerKB, maxFeeNanosPerKB] of the fee range.
+func ComputeTimeBucketRangeFromIndex(index int, feeBucketBaseRateNanos *big.Float, feeBucketMultiplier *big.Float) (
+	_minFeeNanosPerKB uint64, _maxFeeNanosPerKB uint64) {
+
+	minFeeNanosPerKB := computeTimeBucketMinFromIndex(index, feeBucketBaseRateNanos, feeBucketMultiplier)
+	maxFeeNanosPerKB := computeTimeBucketMinFromIndex(index+1, feeBucketBaseRateNanos, feeBucketMultiplier) - 1
+	return minFeeNanosPerKB, maxFeeNanosPerKB
+}
+
+// ComputeTimeBucketIndexFromFeePerKBNanos takes a fee rate, feeBucketBaseRateNanos, and feeBucketMultiplier, and
+// returns the fee range index.
+func ComputeTimeBucketIndexFromFeePerKBNanos(feePerKBNanos uint64, feeBucketBaseRateNanos *big.Float,
+	feeBucketMultiplier *big.Float) int {
+
+	// Compute the fee time bucket index for the fee rate. We can compute the index as follows:
+	// feePerKBNanos = feeBucketBaseRateNanos * feeBucketMultiplier ^ index
+	// log(feePerKBNanos) = log(feeBucketBaseRateNanos) + index * log(feeBucketMultiplier^index)
+	// index = (log(feePerKBNanos) - log(feeBucketBaseRateNanos)) / log(feeBucketMultiplier).
+
+	feeFloat := NewFloat().SetUint64(feePerKBNanos)
+	// If the fee is less than the base rate, return 0.
+	if feeFloat.Cmp(feeBucketBaseRateNanos) < 0 {
+		return 0
+	}
+
+	logFeeFloat := BigFloatLog(feeFloat)
+	logBaseRate := BigFloatLog(feeBucketBaseRateNanos)
+	logMultiplier := BigFloatLog(feeBucketMultiplier)
+	subFee := Sub(logFeeFloat, logBaseRate)
+	divFee := Div(subFee, logMultiplier)
+	feeTimeBucketIndexUint64, _ := divFee.Uint64()
+	feeTimeBucketIndex := int(feeTimeBucketIndexUint64)
+	if feeTimeBucketIndex < 0 {
+		return 0
+	}
+
+	// Now verify that float precision hasn't caused us to be off by one.
+	if computeTimeBucketMinFromIndex(feeTimeBucketIndex,
+		feeBucketBaseRateNanos, feeBucketMultiplier) > feePerKBNanos {
+		// If this condition is true, then the computed index is overestimated by 1.
+		return feeTimeBucketIndex - 1
+	} else if computeTimeBucketMinFromIndex(feeTimeBucketIndex+1,
+		feeBucketBaseRateNanos, feeBucketMultiplier) <= feePerKBNanos {
+		// This condition gets triggered exactly on fee bucket boundaries, i.e. when feePerKBNanos = FeeMax for some bucket.
+		// The float rounding makes the number slightly smaller like 5.9999999991 instead of 6.0.
+		return feeTimeBucketIndex + 1
+	} else {
+		return feeTimeBucketIndex
+	}
```

Comment:
avoid else when possible 😄 maebeam beat this into my head.
```suggestion
	}
	if computeTimeBucketMinFromIndex(feeTimeBucketIndex+1,
		feeBucketBaseRateNanos, feeBucketMultiplier) <= feePerKBNanos {
		// This condition gets triggered exactly on fee bucket boundaries, i.e. when feePerKBNanos = FeeMax for some bucket.
		// The float rounding makes the number slightly smaller like 5.9999999991 instead of 6.0.
		return feeTimeBucketIndex + 1
	}
	return feeTimeBucketIndex
```